crates/hc_seed_bundle/src/seed_cipher/pw_utils.rs

```diff
@@ -38,79 +31,93 @@ where
     let a3 = a3.as_bytes();
 
     // create the output buffer
-    let out =
-        sodoken::BufWrite::new_mem_locked(a1.len() + a2.len() + a3.len())?;
+    let mut out = sodoken::LockedArray::new(a1.len() + a2.len() + a3.len())?;
 
     {
-        // output buffer write lock
-        let mut out = out.write_lock();
+        let mut lock = out.lock();
 
-        // copy / concatonate the three answers
-        out[0..a1.len()].copy_from_slice(a1);
-        out[a1.len()..a1.len() + a2.len()].copy_from_slice(a2);
-        out[a1.len() + a2.len()..a1.len() + a2.len() + a3.len()]
+        // output buffer write lock
+        // copy / concatenate the three answers
+        (&mut *lock)[0..a1.len()].copy_from_slice(a1);
+        (&mut *lock)[a1.len()..a1.len() + a2.len()].copy_from_slice(a2);
+        (&mut *lock)[a1.len() + a2.len()..a1.len() + a2.len() + a3.len()]
             .copy_from_slice(a3);
 
         // we forced utf8 above, so safe to unwrap here
-        let out = std::str::from_utf8_mut(&mut out).unwrap();
+        let out_str = std::str::from_utf8_mut(&mut lock).unwrap();
 
         // this needs a mutable buffer, so we have to do this in out memory
-        out.make_ascii_lowercase();
+        out_str.make_ascii_lowercase();
     }
 
     // return the read-only concatonated passphrase
-    Ok(out.to_read())
+    Ok(out)
 }
 
 /// Use the given passphrase to generate a deterministic secret with argon.
 /// Use that secret to secretstream encrypt the given seed.
 /// Return the argon salt, and the secretstream header and cipher.
 pub(crate) async fn pw_enc(
-    seed: sodoken::BufReadSized<32>,
-    passphrase: sodoken::BufRead,
+    seed: SharedSizedLockedArray<32>,
+    passphrase: SharedLockedArray,
     limits: PwHashLimits,
-) -> SodokenResult<(
-    sodoken::BufReadSized<{ sodoken::hash::argon2id::SALTBYTES }>,
-    sodoken::BufReadSized<24>,
-    sodoken::BufReadSized<49>,
-)> {
+) -> Result<
+    (
+        [u8; sodoken::argon2::ARGON2_ID_SALTBYTES],
+        [u8; 24],
+        [u8; 49],
+    ),
+    OneErr,
+> {
     // pre-hash the passphrase
-    let pw_hash = sodoken::BufWrite::new_mem_locked(64)?;
-    sodoken::hash::blake2b::hash(pw_hash.clone(), passphrase).await?;
-
-    // generate a random salt
-    let salt = sodoken::BufWriteSized::new_no_lock();
-    sodoken::random::bytes_buf(salt.clone()).await?;
+    let mut pw_hash = sodoken::SizedLockedArray::<64>::new()?;
+    sodoken::blake2b::blake2b_hash(
```

Comment:
Is a hash using enough CPU to require a `spawn_blocking`? I'm assuming not but happy to make changes if that's recommended

Reply:
Indeed, unless you are hashing gigs of data it doesn't use up enough cpu realtime to affect task scheduling.

Reply:
LG :+1: